iterfzf/__init__.py

```diff
@@ -54,25 +55,56 @@ def iterfzf(
         cmd.append("--multi")
     if not mouse:
         cmd.append("--no-mouse")
-    if print_query:
-        cmd.append("--print-query")
     if query:
         cmd.append("--query=" + query)
     if preview:
         cmd.append("--preview=" + preview)
+    # always use --print-query: makes things simpler later as return values are always the same
+    cmd.append("--print-query")
+    query_result, *results = _exec_cmd(cmd, iterable, encoding, executable)
+    if results:
+        return_value = results[0:] if multi else results[0]
+    else:
+        return_value = None
+    if print_query:
+        return query_result, return_value
+    else:
+        return return_value
+
+
+def _exec_cmd(cmd, iterable, encoding, executable):
+    """Create subprocess with fzf and return its output to the caller.
+    Most of the work here is related to making sure that the encodings
+    in both directions are correct.
+    In either case returns a tuple where the first element is the query
+    which will be returned if query_result=True as well as the chosen items
+
+    Arguments:
+        cmd List[str] -- Command to be executed via subprocess
+        ... -- just passed through from function call above
+
+    Raises:
+        ValueError: If iterable contains unsuitable elements
+
+    Returns:
+        str,List[str] -- The query (search term entered in fzf) + list of results
+    """
     encoding = encoding or sys.getdefaultencoding()
     proc = None
-    stdin = None
-    byte = None
-    lf = u"\n"
-    cr = u"\r"
-    for line in iterable:
-        if byte is None:
-            byte = isinstance(line, bytes)
-            if byte:
-                lf = b"\n"
-                cr = b"\r"
-        elif isinstance(line, bytes) is not byte:
+    iterator = iter(iterable)
+    first = next(iterator, None)
+    if not first:
```

Comment:
How about detecting an empty iterable by catching `StopIteration` instead of giving the second argument to `next()`?

Reply:
I added a commit that does it that way.